### ISSUE TITLE
Interactive Diagnostics 

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/vscode#64566]
## Problem
JS and TS diagnostic messages can get quite large, especially when dealing with frameworks like Vue or react.   The TS team has worked to significantly improve and shorten diagnostic messages, including replacing complex type information with `...`, however there are still cases where you need to see the full type information. Right now, we have to staticly choose to expand the additional information or not. They would like the user to be able to interactively expand types within a diagnostic

In addition, the overall diagnostic experience could be improved. Two areas identified talking with the TS team:

* Syntax coloring of inline code in diagnostics
* Go to definition in diagnostics

## Refs
#11847

## Desired User Experience
Consider a diagnostic message  `T<...>`. We would like the `...` to be something like a  link. Clicking it would expand the diagnostic message to `T<Array<...>>`, which itself contains another expandable link

## Requirements
The main focus is on being able dynamically expand sections of a diagnostic. We need new VS Code API support for this and then will need languages to adopt this API. However, I think the other two features should be considered while designing the new diagnostics API as they are likely something we would like to support and will influence overall api design 

For expansion, we want to be able to:

- Mark arbitrary spans within a diagnostics as expandable
- Lazily compute these expansions, i.e. don't resolve the expansions ahead of time
- Allow expansions to contain other expansions
- Show a progress indicator of some kind while an expansion is being resolved

**Comment by mjbvz** [comment]
## DiagnosticPart Api sketch
This api sketch proposes making diagnostic messages a list of parts. Parts would be joined together to form the message rendered in the editor:

```ts
interface TextDiagnosticPart {
    readonly type: 'text';

    /**
     * Text rendered in the editor
     */
    readonly text: string;
}

interface ExpansionDiagnosticPart {
    readonly type: 'expansion';

    /**
     * Contents rendered for the expansion, such as the text `...`
     */
    readonly contents: ReaonlyArray<DiagnosticPart>;

    /**
     * Lazily resolves the diagnostics.
     * 
     * The could also be implemented using a provider or commands.
     */
    expand(): Thenable<ReaonlyArray<DiagnosticPart>>;
}

type DiagnosticPart = TextDiagnosticPart | ExpansionDiagnosticPart;
```

The diagnostic type would then become:

```ts
export class Diagnostic {
    ...
	
    message: string | ReadonlyArray<DiagnosticPart>;

     ...
}
```

**Extensions**
The idea of diagnostic parts could be easily extended to clickable links and code:

```ts
interface ClickableDiagnosticPart {
    readonly type: 'clickable';

    /**
     * Contents rendered in the diagnostics 
     */
    readonly contents: ReaonlyArray<DiagnosticPart>;

    /**
     * Command executed on click
     */
    readonly command: Command
}

interface CodeDiagnosticPart {
    readonly type: 'code';

    /**
     * Language to use for syntax highlighting
     */
    readonly language: string

    /**
     * Code rendered in the editor
     */
    readonly text: string
}
```

**Comment by amcasey** [comment]
FYI @minestarks

**Comment by weswigham** [comment]
Probably worth noting that for the syntax highlighting bit, it's very possible that syntax highlighted and linked/expanded ranges overlap; so them being mutually exclusive parts would seem to prevent that.

**Comment by mjbvz** [comment]
@weswigham Good point. I've updated `CodeDiagnosticPart` to potentially handle nesting better. One difficult thing about colorization is that the snippets may not actually be valid code. `T<...>` for example would colorize `T` as a `variable.other.constant` instead of as a type if we reused the js/ts grammar.

In the TS / editor sync today, we also discussed a proposal that would allow specific spans to marked specially, something like:

```ts
interface DiagnosticMessage {
    /**
     * Text rendered in the editor
     */
    readonly text: string;

    /**
     * Ranges within the text that can be expanded.
     */
    readonly expandableRanges?: Range[];

    /**
     * Ranges within the text that should be rendered as code
     */
    readonly codeRanges?: { range: Range, language: string } [];
}
```

Keeping the range correct on expansions could get tricky

**Comment by jrieken** [comment]
Is this inventing our own markup language? How about using markdown as proposed by @mjbvz here: 
https://github.com/Microsoft/vscode/issues/54272

**Comment by weswigham** [comment]
Rendered markdown could be used to implement hyperlinks (though you'd need to ensure file urls with locations work fine in markdown link tags) and if you're lucky syntax highlighting (although what's keeping the editor highlighting and the markdown highlighting consistent?) but I don't think it covers the expandable lazy ranges (the original request) - unless that markdown support goes all the way to full blown html/js support.

**Comment by weswigham** [comment]
It's might also worth noting that for the syntax highlighting part, we'd likely be trying to colorize things in specific contexts, so in addition to the standard markdown language selector, you'd also need to be able to somehow specify a specific tmLanguage rule to start in (instead of the root rule - this way our types get syntax highlighted as types and not statements). So in that regard, getting pre-classified spans (a la quickinfo) back from the language service might be easier than messing with picking apart tmLanguage rulesets; but I'm not the most familiar with them, so I'm not sure.

**Comment by jrieken** [comment]
> unless that markdown support goes all the way to full blown html/js support.

Yeah, good point... We could experiment with commands. We already support command-links and we could have a special command that replaces the link text...

> you'd also need to be able to somehow specify a specific tmLanguage rule to start in (instead of the root rule - this way our types get syntax highlighted

The markdown rendering picks up any language that's registered (by extensions). So it would require some special code-chunk-language so that MD will use the grammar that came with that language. 

Stepping back a bit: The idea is to keep of all this in the editor hover, right? So make that smarter and a bit more interactive? I am asking because the hover is easy to dismiss and maybe not the best place for a UX in which the user invests time configuring its state.

**Comment by sandy081** [comment]
My 2 cents, I would not have asynchronosity in error message which might not have good UX especially in Hovers. Why does this needs to be async first of all Is it expensive to compute complete message?

Also not a big fan of having UI semantics. My first step would be to use MD string. BTW, Is not the related information can be used to show more information? I think we do not restrict this to be used only for multi-diagnostics but also for adding more information.

**Comment by weswigham** [comment]
> Why does this needs to be async first of all Is it expensive to compute complete message?

It can be, yes. Our typesystem has the ability to generate new types by mapping old ones, so you can get very complex or large (anonymous) types with a very short sequence of type operators. And when those appear in errors, well... Things can get difficult to grok very quickly since there's simply too much to print - today we're forced to truncate the output, but making it explorable just like how dotting off the object can reveal more info would go a long way toward making it easier to work with.

> My 2 cents, I would not have asynchronosity in error message which might not have good UX especially in Hovers.

Ignoring hover over red squiggles for a moment, errors still appear in the problems window and the f8 error bar as well, both of which are much more persistent. I'd think persisting the hover text as it is updated is an extension of those two.

> BTW, Is not the related information can be used to show more information?

Nope - in this case the lazy querying is key to making it work. It allows us to gradually expose/generate the shape of complex types as requested, and only the parts requested. We _actually_ have the ability to construct infinitely deep anonymous types - today we cut these off after printing one level of the type with a `...`, but being able to expand these to an arbitrary depth as needed would be very nice. We do use related info where we can - it just doesn't help in complex type display/type truncation cases. (And secondarily would be a gratuitious way to "linkify" all the references in the types in an error, TBH - very nonlocal)

> Stepping back a bit: The idea is to keep of all this in the editor hover, right? So make that smarter and a bit more interactive? I am asking because the hover is easy to dismiss and maybe not the best place for a UX in which the user invests time configuring its state.

What I was hoping was that diagnostic state was synchronized between the problems window/the f8 error bars/hover strings between full refreshes. That way you _can_ expand within the hover if you want (then go back to it and still see it expanded), but also explore in the other places errors are displayed. In our meeting, someone mentioned that eclipse has the ability to copy things like individual diagnostics (plus stack traces and more) into a project context pane to allow them to persist and be inspected, which could work, too, but is maybe less continuous an experience than just being able to interact with a diagnostic anywhere you see it.

**Comment by jrieken** [comment]
adding @misolori - call for diagnostics UX ideas

**Comment by mjbvz** [comment]
Here's the most minimal API for this we could have on the vscode side:

1. Make `Diagnostic::message` a markdown string
1. In the diagnostic messages, allow hyperlinks of the form:

    ```md
    some text [...](command:editor.expandDiagnostic?_typescript.resolveDiagnostic,123) and more text
    ```

1. The `editor.expandDiagnostic` command is a special VS Code command that delegates out to another command to handle the actual expansion. In this case, it would invoke the `_typescript.resolveDiagnostic` command with the diagnostic object and the expansion id `123`

1.  `_typescript.resolveDiagnostic` would return a promise to a `MarkdownString`. VS Code would then replace the original `editor.expandDiagnostic` link with the rendered string

A few things with this approach:

1. This would let you use other markdown features like bolds and italics. 
1. Can we improve this so we don't need the command indirection? 
1. Commands don't have good cancellation semantics
1. This is pretty VS Code specific. TypeScript would still likely want a diagnostic structure api closer to the one I originally proposed. 
1. This approach doesn't address inline syntax highlighting. That's not possible to do in markdown as far as I know

**Comment by jrieken** [comment]
> _typescript.resolveDiagnostic would return a promise to a MarkdownString. VS Code would then replace the original editor.expandDiagnostic link with the rendered string

In case TypeScript already knows the omitted text behind `[...]` it could just add it as argument, `[...](command:md.expand?FOO`). That would prevent extra commands and round trips. 

> This would let you use other markdown features like bolds and italics.

Yes. My POV is that it's just like `Hover` and editor decorations and I believe people can build cool things with it, see GitLens. Sure, there is a risk of folks turning us into a Christmas tree but then I don't use such an extension. We can also consider disabling certain markdown features, e.g. tables or images, in certain places.

> This is pretty VS Code specific. TypeScript would still likely want a diagnostic structure api closer to the one I originally proposed.

Haha - better VS Code specific extension API than TypeScript specific extension API. Honestly, the proposal smells very TypeScript-tailored and we should first circle in other languages with complex error messages, maybe rust-lang or lean.

**Comment by Ciantic** [comment]
I think there should be a way to automatically expand all of the three dots when selecting the text, I use this shortcut:

`Ctrl+K Ctrl+I` Hover, `Ctrl + A` select all `Ctrl+C` copy

And when I select all I want the whole type, not the abbreviated form.

**Comment by Arlen22** [comment]
Please allow these to be expanded somehow. It has become VERY frustrating working with types now in typescript. I'm wasting so much time trying to get everything to work with the different intellisense features, and not being able to copy and paste complete types, or even read them properly has become VERY frustrating and time consuming. PLEASE PLEASE PLEASE add SOME way (even just a command) to show the full type without any abbreviations. 

One problem is that I don't even know which type is being repeated or if more than one is being repeated. A type variable layout would be more clear. 

I'm used to hacking other peoples code, so I am perfectly capable of building an extension or changing the source code, I just don't want to have to do that with everything I use. Why can't the tools just do what they're supposed to do and get out of the way?

**Comment by Arlen22** [comment]
I see at the top that this can be staticly chosen, but I have no idea how to do that. Is it a setting?

**Comment by asanacam** [comment]
Just checking in on this. Is there any update to this, or any way to get rid of the ellipses and see the full type of a variable in VS Code?

**Comment by taxilian** [comment]
This is a really annoying issue; honestly there is never a time when I don't want to see the whole thing. Optimize for readability? How is it readable to only see a small part of the relevant information?

Would really really like to see this become a setting that was easy to tweak :-( it would save me a lot of time right now with what I'm working on

**Comment by Arlen22** [comment]
## Problem solved!
### `noErrorTruncation: true`
At some point someone pointed out the TypeScript tsconfig.json setting `compilerOptions.noErrorTruncation`. This in fact seems to solve the problem.

**Comment by Ciantic** [comment]
@Arlen22 It doesn't work.

![image](https://user-images.githubusercontent.com/64731/75149139-f41bb480-5709-11ea-80c0-763d7dc90f11.png)

<details>

```typescript
// from https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/intro-to-tweet-json
let foo = {
  "text": "My added comments to this Tweet ---> https:\/\/t.co\/LinkToTweet",
  "user": {
    "screen_name": "TweetQuoter"
  },
  "quoted_status": {
    "text": "original message",
    "user": {
      "screen_name": "OriginalTweeter"
    },
    "place": {      
    },
    "entities": {      
    },
    "extended_entities": {      
    }
  },
  "place": {    
  },
    "entities": {
      "hashtags": [

      ],
      "urls": [
        {
          "url": "https://t.co/W1FqVrsjQk",
          "expanded_url": "https://twittercommunity.com/t/rt-with-comment-aka-quote-tweets-now-may-have-media-attached-across-twitter-apis/125539",
          "display_url": "twittercommunity.com/t/rt-with-comm…",
          "unwound": {
            "url": "https://twittercommunity.com/t/rt-with-comment-aka-quote-tweets-now-may-have-media-attached-across-twitter-apis/125539",
            "status": 200,
            "title": "RT with comment, aka Quote Tweets, now may have media attached across Twitter APIs",
            "description": "Today we’re announcing a new feature that allows userspeople on Twitter to add a GIF, video, or photo to their Retweet with comments, aka Quote Tweets. Developers will start to see additional media metadata included in the payload for Quote Tweets. Quote Tweets with media will be rendered very similarly to Tweets with media across our APIs. This rendering is an addition to the existing media field, so this should not be a breaking change for application owners, if they are already ingesting med..."
          },
          "indices": [
            162,
            185
          ]
        },
        {
          "url": "https://t.co/lXs26mCkIl",
          "expanded_url": "https://twitter.com/TwitterSupport/status/1125479034513645569",
          "display_url": "twitter.com/TwitterSupport…",
          "indices": [
            186,
            209
          ]
        }
      ],
      "user_mentions": [

      ],
      "symbols": [

      ],
      "media": [
        {
          "id": 1125490782532657153,
          "id_str": "1125490782532657153",
          "indices": [
            210,
            233
          ],
          "media_url": "http://pbs.twimg.com/tweet_video_thumb/D56L51LV4AEEC8-.jpg",
          "media_url_https": "https://pbs.twimg.com/tweet_video_thumb/D56L51LV4AEEC8-.jpg",
          "url": "https://t.co/T9MBCHZWcD",
          "display_url": "pic.twitter.com/T9MBCHZWcD",
          "expanded_url": "https://twitter.com/TwitterDev/status/1125490788736032770/photo/1",
          "type": "animated_gif",
          "video_info": {
            "aspect_ratio": [
              101,
              165
            ],
            "variants": [
              {
                "bitrate": 0,
                "content_type": "video/mp4",
                "url": "https://video.twimg.com/tweet_video/D56L51LV4AEEC8-.mp4"
              }
            ]
          },
          "sizes": {
            "thumb": {
              "w": 150,
              "h": 150,
              "resize": "crop"
            },
            "small": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            },
            "large": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            },
            "medium": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            }
          }
        }
      ]
    },
    "extended_entities": {
      "media": [
        {
          "id": 1125490782532657153,
          "id_str": "1125490782532657153",
          "indices": [
            210,
            233
          ],
          "media_url": "http://pbs.twimg.com/tweet_video_thumb/D56L51LV4AEEC8-.jpg",
          "media_url_https": "https://pbs.twimg.com/tweet_video_thumb/D56L51LV4AEEC8-.jpg",
          "url": "https://t.co/T9MBCHZWcD",
          "display_url": "pic.twitter.com/T9MBCHZWcD",
          "expanded_url": "https://twitter.com/TwitterDev/status/1125490788736032770/photo/1",
          "type": "animated_gif",
          "video_info": {
            "aspect_ratio": [
              101,
              165
            ],
            "variants": [
              {
                "bitrate": 0,
                "content_type": "video/mp4",
                "url": "https://video.twimg.com/tweet_video/D56L51LV4AEEC8-.mp4"
              }
            ]
          },
          "sizes": {
            "thumb": {
              "w": 150,
              "h": 150,
              "resize": "crop"
            },
            "small": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            },
            "large": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            },
            "medium": {
              "w": 202,
              "h": 330,
              "resize": "fit"
            }
          }
        }
      ]
    }
  };
```

</details>

The bug to track is this: https://github.com/microsoft/TypeScript/issues/26238

**Comment by reggi** [comment]
I am also looking for a way to access "computed types" I've posted on [StackOverflow about it here](https://stackoverflow.com/questions/63044623/create-export-of-computed-types-in-file). It would be nice to have better tooling / understanding about what a type is, and I'd personally like a CLI tool that could give me a glimpse, rather then trying to get VSCode to show everything in a tooltip on hover.

**Comment by lake2** [comment]
same issue + 1

**Comment by plaa** [comment]
Somewhat related is my recently-closed-as-out-of-scope suggestion of [being able to cmd-click on types in the tooltip](https://github.com/microsoft/vscode/issues/97704) to dig into those types (like you can in Eclipse).  That request is also about having "links" within the tooltip.

**Comment by sharedrory** [comment]
I found https://github.com/intellij-rust/intellij-rust/pull/4217 through https://github.com/microsoft/vscode/issues/16221#issuecomment-518658249 and it seems to be doing almost the same thing (it expands all the way with just 1 click). I would imagine it being similar to that except having the option to advance 1 step at a time?

**Comment by danielo515** [comment]
What I'm doing meanwhile is to use [simplify from type fest to wrap big types](https://github.com/sindresorhus/type-fest#:~:text=Simplify%20%2D%20Useful%20to%20flatten%20the%20type%20output%20to%20improve%20type%20hints%20shown%20in%20editors.%20And%20also%20to%20transform%20an%20interface%20into%20a%20type%20to%20aide%20with%20assignability.)

It successfully expands the type on VSCode and I can freely scroll it.

**Comment by djmisterjon** [comment]
am also look a way to get something like on hover
![image](https://user-images.githubusercontent.com/24865815/156882502-8bba026f-7909-493c-9ea6-d3a098d3c4a7.png)
instead of this 
![image](https://user-images.githubusercontent.com/24865815/156882525-38922f20-14da-4a92-82e1-2ee4572a334c.png)

**Comment by zen0wu** [comment]
can we prioritize this, we eagerly need this feature to expand all the complex types in our codebase.

i'm currently doing https://github.com/Microsoft/TypeScript/issues/26238#issuecomment-672086446, but that's undesired because sometimes i don't want to expand the full type.

**Comment by prokopschield** [comment]
I'd just like to +1 that this feature would be much appreciated.

**Comment by 205g0** [comment]
@weswigham @jrieken any news on this issue?

**Comment by dreit-p** [comment]
We have been waiting for any solution for 5 years now, let us know are you considering it? This problem is really annoying and limiting us

**Comment by starball5** [comment]
Related: #94679 and #35601.

Related on Stack Overflow:

- [How can I make VS Code show full TypeScript type definitions on mouse hover?](https://stackoverflow.com/q/62508909/11107541)
- [Is there a way to access typings of a prop in VS Code without having to check the types files?](https://stackoverflow.com/q/76138599/11107541)

**Comment by bombillazo** [comment]
Right now it's like hanging a steak in front of a canine; as a dev seeing the inferred type in the tooltip and knowing VS Code can expose the complete type yet not being able to copy or extract it in any way... This would be the most effective feature to help us avoid having to manually setup up types from a JSON payload or code-gen'ing our types, especially for those that are combinations of other types.

**Comment by aprilmintacpineda** [comment]
It would be nice if we can click on the ellipses or `more` then we can see some more.

**Comment by nabilfreeman** [comment]
This is actually not a VSCode problem but a TypeScript problem.

Proof:

1) Create a very long type that displays `... 15 more ...` when you hover it
2) Deliberately mis-assign it, for example `const foo: VeryLongType = {}` missing required properties or something like that
3) Run `npx tsc`

You'll see even in your terminal TypeScript collapses the type and outputs `... 15 more ...`.

But I would also like to see this fixed, because I often need to copy some complex computed types from one project to another, and this would save me a lot of chore work.

**Comment by vennson** [comment]
Long term solution by [@extremegf](https://github.com/microsoft/TypeScript/issues/35601#issuecomment-1824306948)

in vscode
1. Using Ctrl+P, Go to `node_modules/typescript/lib/tsserver.js`
2. Find line `var defaultMaximumTruncationLength = 160`
3. Change it to `1000` or value of your choosing.
4. Open Ctrl+Shift+P menu, select TypeScript: Select TypeScript Version...
5. Ensure it's set to Use Workspace Version
6. Full restart VSCode - reloading workspace is not enough.
7. Verify that ... abbreviations are gone from type hints.

PS.
cause `noErrorTruncation` in tsconfig didn't work for me.
and changing tsserver.js of VScode's typescript is anoying every VSCode update

Before:
with ...
<img width="342" alt="image" src="https://github.com/microsoft/vscode/assets/28529262/55a53baf-011e-439d-a358-fa853e0c3e81">


After:
You can also drag the bottom border
<img width="345" alt="image" src="https://github.com/microsoft/vscode/assets/28529262/40cb26bb-d402-46fc-a1d6-dbde143f483c">

**Comment by boylett** [comment]
Have written a quick patcher that I run using yarn like so:
`yarn patch-ts`

Add the following to `scripts` in `package.json`:
```json
  "patch-ts": "./scripts/patch-ts.sh",
```

Add the following to `scripts/patch-ts.sh` (and remember to run `chmod +x ./scripts/patch-ts.sh`
```sh
#!/bin/bash

FROM="var defaultMaximumTruncationLength = 160;"
TO="var defaultMaximumTruncationLength = 1e6;"

sed -i -e "s/$FROM/$TO/g" "node_modules/typescript/lib/typescript.js"

echo "Annotation Truncation Length patched successfully"
```

**Comment by BoscoDomingo** [comment]
> Long term solution by [@extremegf](https://github.com/microsoft/TypeScript/issues/35601#issuecomment-1824306948)
> 
> in vscode
> 
>     1. Using Ctrl+P, Go to `node_modules/typescript/lib/tsserver.js`
> 
>     2. Find line `var defaultMaximumTruncationLength = 160`
> 
>     3. Change it to `1000` or value of your choosing.
> 
>     4. Open Ctrl+Shift+P menu, select TypeScript: Select TypeScript Version...
> 
>     5. Ensure it's set to Use Workspace Version
> 
>     6. Full restart VSCode - reloading workspace is not enough.
> 
>     7. Verify that ... abbreviations are gone from type hints.
> 
> 
> PS.
> 
>     * cause `noErrorTruncation` in tsconfig didn't work for me.
> 
>     * and changing tsserver.js of VScode's typescript is anoying every VSCode update
> 
>     * the extension mentioned in other solution is no copy-pasteable
> 
> 
> Before: with ... <img alt="image" width="342" src="https://private-user-images.githubusercontent.com/28529262/344517259-55a53baf-011e-439d-a358-fa853e0c3e81.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczMzk1MzcsIm5iZiI6MTcyNzMzOTIzNywicGF0aCI6Ii8yODUyOTI2Mi8zNDQ1MTcyNTktNTVhNTNiYWYtMDExZS00MzlkLWEzNTgtZmE4NTNlMGMzZTgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDA4MjcxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhYjQ4YTI5N2QzYjNkZTU0YjliOTlkMWQzNjljZGU0MGVlYWRjMzEwZDI0MjM0Njk5ZjI4MDgyNTg4NTMzNGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8csN9mbAhUQ267Cu4BYOQayNmbSfi8PRCpa7tD32oQ8">
> 
> After: You can also drag the bottom border <img alt="image" width="345" src="https://private-user-images.githubusercontent.com/28529262/344517136-40cb26bb-d402-46fc-a1d6-dbde143f483c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczMzk1MzcsIm5iZiI6MTcyNzMzOTIzNywicGF0aCI6Ii8yODUyOTI2Mi8zNDQ1MTcxMzYtNDBjYjI2YmItZDQwMi00NmZjLWExZDYtZGJkZTE0M2Y0ODNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDA4MjcxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjMzZlNDM2YjhiMjVkNWM4NmIwZmFlN2ZmZGM1ODAxMzZiNjI3YTBlMjZhMmUxNjVmZjEyNzdmYzkzOTA2MDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XOKsj-EI-OQMvfEfQu9FDaeUG4fz4-IRtno1DzhEsMo">

Since TypeScript 5.6 this is now located in `node_modules/typescript/lib/typescript.js`

**Comment by WendaoLee** [comment]
> Have written a quick patcher that runs automatically after installing new dependencies or updates using a [post script](https://docs.npmjs.com/cli/v10/using-npm/scripts#pre--post-scripts).已经编写了一个快速修补程序，它会在使用[后置脚本](https://docs.npmjs.com/cli/v10/using-npm/scripts#pre--post-scripts)安装新依赖项或更新后自动运行。
> 
> 1. Add the following to `scripts` in `package.json`:将以下内容添加到 `package.json` 中的 `scripts`：
> 
>   "postinstall": "./scripts/patch-ts.sh",
> 2. Add the following to `scripts/patch-ts.sh` (and remember to run `chmod +x ./scripts/patch-ts.sh`将以下内容添加到 `scripts/patch-ts.sh` 中（并记得运行 `chmod +x ./scripts/patch-ts.sh`
> 
> #!/bin/bash
> 
> FROM="var defaultMaximumTruncationLength = 160;"
> TO="var defaultMaximumTruncationLength = 1e6;"
> 
> sed -i -e "s/$FROM/$TO/g" "node_modules/typescript/lib/typescript.js"
> 
> echo "Annotation Truncation Length patched successfully"
> 3. Run `yarn install` or `npm install`运行 `yarn install` 或 `npm install`
> 4. Reload Window  重新加载窗口
> 5. ???
> 6. Profit  利润

For windows user,here is `powershell` script code:

```ps1
# Only for typescript 5.5.2 above
$FROM = 'var defaultMaximumTruncationLength = 160;'
$TO = 'var defaultMaximumTruncationLength = 1e6;'
$FILE_PATH = "node_modules/typescript/lib/typescript.js"

$content = Get-Content -Path $FILE_PATH -Raw

$newContent = $content.Replace($FROM, $TO)

$newContent | Set-Content -Path $FILE_PATH -NoNewline

Write-Host "Annotation Truncation Length patched successfully" -ForegroundColor Green

```

**Comment by trusktr** [comment]
> - Go to definition in diagnostics

Can the click-on-type-inside-tooltip-to-go-to-definition be split to a separate GitHub ticket and implemented separately?

That seems like a smaller task than this overall ticket, and would be really helpful on its own.

I already opened this one,

- https://github.com/microsoft/vscode/issues/109264 (or Go To Definition, at least)

but that one got rolled into this one.
